dartagnan/src/main/java/com/dat3m/dartagnan/program/event/EventFactory.java

```diff
@@ -602,4 +602,45 @@ public static Fence newLwSyncBarrier() {
         }
     }
 
-}
+    // =============================================================================================
+    // ============================================ PTX ============================================
+    // =============================================================================================
+    public static class PTX {
+        private PTX() {}
+
+        public static Store newTaggedStore(IExpr address, ExprInterface value, String sem,  String scope) {
+            Store store = new Store(address, value, sem); // sem = REL || RLX || WEAK
+            store.addFilters(scope); // scope = CTA || GPU || SYS
+            return store;
+        }
+
+        public static Load newTaggedLoad(Register register, IExpr address, String sem, String scope) {
+            Load load = new Load(register, address, sem); // sem = ACQ || RLX || WEAK
+            load.addFilters(scope); // scope =  CTA || GPU || SYS
+            return load;
+        }
+
+        public static Fence newTaggedFence(String sem, String scope) {
+            Fence fence = new Fence(sem); // sem = ACQ_REL || SC
+            fence.addFilters(scope);
+            return fence;
+        }
+
+        public static RMWFetchOp newTaggedAtomOp(IExpr address, Register register, IExpr value,
+                                           IOpBin op, String sem, String scope) {
+            RMWFetchOp atom = new RMWFetchOp(address, register, value, op, Tag.Linux.MO_MB);
+            atom.addFilters(scope);
+            atom.addFilters(sem); // sem = ACQ_REL || RLX
+            return atom;
+        }
+
+        public static RMWOp newTaggedRedOp(IExpr address, Register register, IExpr value,
+                                                    IOpBin op, String sem, String scope) {
+            RMWOp red = new RMWOp(address, register, value, op);
```

Comment:
Sorry I don't get how should I update this part.

Reply:
You added two event factory methods that create RMW events for PTX: `newTaggedAtomOp` and `newTaggedRedOp`. Those methods returns events `RMWFetchOp` and  `RMWOp`. However, you can see that the events are in the `event.lang.linux` package, showing that they are intended to be used with the Linux kernel memory model (LKMM). Also, you can see in those classes that these events use LKMM specific tags, e.g.,
```
    public RMWOp(IExpr address, Register register, IExpr value, IOpBin op) {
        super(address, register, value, Tag.Linux.MO_ONCE);
        this.op = op;
        addFilters(Tag.Linux.NORETURN);
    }
```
What you need to do is create a new package `event.arch.ptx` and two new events `AtomOp` and `RedOp`. Then the event factory would create instances of these events. Since now both the LKMM and the PTX events would extend `RMWAbstract`, you would need to move these abstract class to `event.core`.

Once you introduce this new method, you will also need to modify `program.processing.compilation.VisitorPTX` accordingly

Reply:
Updated.